### ISSUE TITLE
API6 enhancements

### ISSUE THREAD

**Issue by izyspania** [ampache/ampache#3607]
I have some enhancements requests for API6 that exists in subsonic API and they are pretty useful (at least for apps)

1.Bookmarks API should include Album / Artists and more info about the song, i have to do too extra requests to display a list of bookmarks if i want to include the arts and song duration, artists info, get the play URL etc. in the "Bookmarks tab / page".
2. Stats API should include an "alphabetical by name" (maybe alphabetical by artist too for albums) option like it does with subsonic API , at the moment i dont know how to get them in alphabetic order by API and its a mess of combining multiple API requests with internal sorting as im using some offsets or sizes because i dont want to load all the albums / songs / artists at once (im loading them on scroll in my app).
3.Flagged API (i know now that we can use Stats API but..) - would be nice to have a separate API like Bookmarks API for this that returns albums / artists and songs in the same request,
4.Forgot password API (there is an open topic on this but i thought i should include it here in the list)
5.Image size in the API (there is an open topic on this but i thought i should include it here in the list)
6 .Can we use advanced_search to search artists / albums / songs with a single request? at the first look i didnt find a way , would be great to have it , even a normal Search API like "search2" is in Subsonic API would be great

Thanks,

**Comment by izyspania** [comment]
Any update on this ? Are you considering adding any of this?

Also the API bookmark_delete ( action=bookmark_delete ) doesnt seem to work , it returns: 
{
    "bookmark": []
}
and no action is taken.

Some questions:
Maybe bookmark_delete  should use the bookmark id also instead of objectid?
I also noticed that album/song/ artist etc. ids are not unique (they were using subsonic) , is that intended? (like a song can have the same id with an album etc. , its not a problem but yeah)

**Comment by lachlan-00** [comment]
id's are never unique, only unique per type.

I'm working on lost_password now and i'll keep extending these bookmarks out to use the id as well. just haven't got large blocks of time lately

**Comment by lachlan-00** [comment]
added lost_pasword method in 40b56f5ba

create a reset token using the username and email to send a reset token

**Comment by izyspania** [comment]
Great , thanks a lot for your work, lost_pasword is a great add.

As an update:
Im using an internal sorting in the app for "Albums and Artists API" at the moment and it works fine but the problem is that if you have a lot of artists / albums and you want to use pagination so you can limit the amount of albums / artists loaded at once (with offset and limit) you cant really sort them and would be great if there was an option for some basic sorting.

If some basic sorting can be added to Albums / Artists etc i think Stats API doesnt need to have more sorting options (maybe only a "by name") but it should have an "include" option , its a great performance upgrade because i can just pass the "songs, artits info etc.." instead of querying the album / artists again when you want to go from the List to a specific album/artist etc.

Bookmarks API should really have "include" option or just include the artists/albums by default , not sure if it makes sense for general use case but for my particular use i have to do many queries to display the bookmarks info then play them (bookmarks have an important role in my project)
Example:   Query Bookmarks -> Query Song with the id from Bookmarks (to get info about the track / album so i can display track/song total time, art, info about the Album (book) etc) -> (on play) query Album with the id from the Song Query (to get all the tracks from the current album , the idea of is this is that when i want to play a bookmark i want to add all the tracks of the Album in a Que so it can play the next track after the current is finished , atm each Album is a Book with Chapters in my APP , so when i play a Chapter from a Book im adding all the "tracks/chapters" in the Que if it makes sense) , it would be great to be able to skip the middle query.

**Comment by lachlan-00** [comment]
all the X's methods sort by name wihtout other options  right now so if you want different sorting that doesn't work
![image](https://github.com/ampache/ampache/assets/1305249/d8613a5a-e1eb-467d-beb1-30e24fd40aef)

But all we have to do is extend to allow sorting on the browse.
![image](https://github.com/ampache/ampache/assets/1305249/e9c04a84-5db4-4e59-9a57-6693b91c80a3)

for bookmarks i'll be including the object itself as a sub item but haven't got to that one yet

**Comment by lachlan-00** [comment]
7a5a2535d add bookmark as a type for edit/delete

0644b1ec1 add bookmark to get a dingle bookmark by id

**Comment by izyspania** [comment]
Thats great thanks , ill test the changes later , do you update the develop demo server after each code update?

**Comment by izyspania** [comment]
The song URL info comes like this:
 "url": "https:\/\/develop.ampache.dev\/play\/index.php?ssid=7e951b11262f067dffd7e618896331bc&type=song&oid=185&uid=3&player=api&name=Nofi-found.%20-%20Invo.mp3",

I see that the ssid = auth, what is it different?
In my my app im building the stream link myself for each song instead of using the URL , wasnt it better to be only like "&type=song&oid=185&uid=3&player=api&name=Nofi-found.%20-%20Invo.mp3" instead of the full link? The auth key could expire and if you cache the song info you wont be able to access the file or im missing something here?

**Comment by lachlan-00** [comment]
Basically because it's always been like that.

The api came long after song streaming so i assume they just kept it as is for back compatibility. The original idea is that you log in and then the play urls are generated on demand. Things like ip and location can dictate different urls and all the different configs could change parameters based on many things.

To make your own links permanent, users have streaming tokens available which never expire.
so if you want permanent links for your urls give all your users a streaming token.
https://github.com/ampache/ampache/wiki/ampache6-details#allow-permalink-user-streams

a streaming token can only stream for your user but it will never expire until rotated. it's available in the user responses

**Comment by lachlan-00** [comment]
I've updated lost_password to use auth instead of token which allows putting it in the header. https://ampache.org/api/#http-header-authentication

it's also documented now https://ampache.org/api/api-json-methods#lost_password

stream urls will probably have updates for header auth as well but i haven't got to that yet

**Comment by izyspania** [comment]
Yeah , i know , its just like the Api Key but only for streams (very cool) but the problem with both of them (Api Key and stream token) is that i have to manually create one for each user , isnt it possible to add an option in the panel that will auto create Api Key / Stream token on registration?

Also , if i understood correctly , if i give a user a stream token it will replace the token in the url from song info etc? If not how is it possible to get the streaming token with the API ? 

Would be great to include savePlayQueue , getPlayQueue and NowPlaying APIs (they exist in the subsonic API, NowPlaying  could be included in the Stats method)

Im still working on replacing the API from subsonic to ampache in my app and im only using the handshake with an API KEY for auth at the moment and i re-auth when the session expires, still struggling to find the best way to handle the auth in my app.

The main problem is the expired session in 2 different ways , one is if you idle in the app (or app in background) and the session expires for most of the api calls i can re-auth and use the new auth key but if you had a song(book)/podcast etc on pause then i have to re-build the que with all the album songs so i can add the stream url with the new keys. 
My options now seem to be:
 1) to PING the auth key at an X amount of time so i can keep the session alive (its a bit tricky to do this on android on how android handles the background stuff, it mostly pauses everything) 
 2) handle the error on Play/Pause and re-auth then rebuild the que with the new key (its a bit tricky to do this also because if you have the app in background and you control the play from your notification area not sure if i can rebuild the que with the app in background , ill have to dig more into this)
3) Is it possible to make the auth key live forever / for a longer period of time (or until goodbye) ? Is it possible to have only one session per user too (like invalidate old auth if a new handshake was performed) ?
4) Possible to get the API KEY / STREAM key on handshake? (if we could have auto generated api keys this would be useful)

Using stream key seems like a flaw as they can be grabbed and shared after but seems that it can be an option.

Ill have to test lost_password API, bit busy with everything else.

**Comment by lachlan-00** [comment]
i'll expand more when i'm not working but.

stream token is going to be auto per user at some point. i'll look at options for generating both or disabling on create in config. (easy to do)

streaming token is returned in a user object.
![image](https://github.com/ampache/ampache/assets/1305249/1b557798-b049-4c6a-8578-e9508590ae25)

It might be good to return the streamtoken on handshake or ping too (also easy todo)

You can allow a stream to never die the session expiry can be set with session_length. i might look at something similiar  to remember_length for the api as well. give people a 1 week session by using remember_length instead of session_length. (setting either of those really high should always extend with use.)

I will check that setting the session_length to 0 will just keep them forever

To allow permanent auth tokens i would probably only allow that in the header. how... not sure but i wouldn't let them be sent in parameters. that needs more work as it's generally expected that sessions retire at some point.

so semi-permanent auth with session_length and remember_length almost do it but i want to work on a permanent option using header tokens.

**brain dump part below**

my thought on that is the server allows an app as admin (eg admin runs a method like set_token or runs a cli command to generate a shared token) and then the app must send some kind of variation of this token with their request.

So an example python app will have a constant that i use to hash the user and their password/token which will be verified by the server using that app token as a salt but it's not sent by the app.

This can be in the app or updated by connecting to the server with the user and password with some kind of app verification against the client string. not sure but it's a larger piece of work than extending sessions

**Comment by izyspania** [comment]
For now i was just thinking to extend the session and re-auth when needed , now im requestion a new auth key if the server responds with "expired session" but maybe i should update based on session_expire, also im thinking to send a PING each X amount of time if they have the song paused so it will have a valid session when re-play or just check for the session_expire on play (for this ill need to rebuild the que and im not a fan of it but its an OK solution) .

Now im working on the other things but when ill get to the auth part ill see what will be the best solution also would be great if there was a way to allow only specific apps to access the server but this is a bit tricky i guess, one simple option would be have a salt in the app but i guess someone could just crack the app and get it but it would still be better than nothing. (if ill get some idea for something simple ill post)

**Comment by izyspania** [comment]
I am trying to implement the search now but i ran in a few issues (more or less).

Got a page that searches for albums/artists/songs at the same time , my first idea was using the albums/artists/songs methods as they are more simple to use but when its empty they return an error instead of [] , wasnt it better to return [] like the advanced_search does? Problem with returning an error is that  i have to make some structural changes just for this in the way im handling errors now.

Getting this:
{
    "error": {
        "errorCode": "4704",
        "errorAction": "albums",
        "errorType": "empty",
        "errorMessage": "No Results"
    }
}
was expecting this:
{
    "album": []
}

I am looking on using advanced_search instead as it returns [] for empty instead an error but it would be great if you could use multiple types like &type=artists,songs,albums& if possible.

**Comment by izyspania** [comment]
I see that compared with subsonic API the Song object is missing "duration" info, would be great if you can add that also.

**Comment by lachlan-00** [comment]
> I see that compared with subsonic API the Song object is missing "duration" info, would be great if you can add that also.

time is the song time/duration/length in seconds.

**Comment by lachlan-00** [comment]
> I am trying to implement the search now but i ran in a few issues (more or less).
> 
> Got a page that searches for albums/artists/songs at the same time , my first idea was using the albums/artists/songs methods as they are more simple to use but when its empty they return an error instead of [] , wasnt it better to return [] like the advanced_search does? Problem with returning an error is that i have to make some structural changes just for this in the way im handling errors now.

Albums is done using the (future todo) api handling and raises an exception when empty.

artists and songs return empty so for albums that's a bug to fix, it should return empty the same as the others. i'll do a run on the %object%'s methods and they should return empty object arrays

**Comment by izyspania** [comment]
> time is the song time/duration/length in seconds.

Ah sorry i missed the "time" for some reason.

**Comment by lachlan-00** [comment]
I've checked all the %'s methods and they should all return empty results.

searching multiple objects isn't possible right now without extending. but i will be extending search into 'collections' which will allow any type of object to be returned.

It will be there eventually but needs a structural change and table updates/new ones.

**Comment by izyspania** [comment]
> I've checked all the %'s methods and they should all return empty results.
> 
> searching multiple objects isn't possible right now without extending. but i will be extending search into 'collections' which will allow any type of object to be returned.
> 
> It will be there eventually but needs a structural change and table updates/new ones.

Great , thanks for your work , im just using a local search with the cached data from getting all albums/artists at the moment, i am thinking that this will have better performance in the app (everyone has a decent phone nowdays) and will help with performance on the server side too but im not sure if it will stay like this or ill get back using the search API, ill have to see when im done with everything how many albums/artists ill use , if there are too many ill have to use the search if not i think using cached data is better.

**Comment by izyspania** [comment]
I am still struggling to find the best way to use the auth in app, the problem is when the key expires, at the moment im requesting a new key but i have cashed things so they will still use the old auth key, ok for them i can just update the key even its not the best way to do it but problem is when i use multiple queries in parallel (like on the "home page", using stats API ) , being at the same time each one of them will get "session expired" and will request a new key , a bit of a mess , i can work around this but it just doesnt feel right somehow.

If you can do something about this (or if im doing something wrong and you have some idea) would be nice , like Subsonic API only uses the hashed password , maybe you can add it to the ampache API too or just have a separate permanent auth key instead of the current auth key that we can  use for APPs that  will be based on the user password (if user changes his password then this new auth key will change) that would make things way more simpler, apps dont really need variable auth keys that can expire. If its not possible or too much work for it i guess ill have to work with managing sessions but at least if it would be possible to extend an expired key, that would be a great solution too.

**Comment by lachlan-00** [comment]
i'm putting a permanent auth in the header soon using the bearer token https://ampache.org/api/#http-header-authentication

Just need to work it out and test

**Comment by lachlan-00** [comment]
what are some of the things that are cached with a session auth that you're replacing?

**Comment by lachlan-00** [comment]
With b5a8f9f4e i've made the initial continuous auth changes.

When you auth with a header token you will get a simple session (repeatable `md5(username)`)
![image](https://github.com/ampache/ampache/assets/1305249/f90d35ab-de31-4fb1-89ca-78dd799d69d4)

The session key will always be the same but it requires you to perform actions using the original header token

so if you use the simple session as a url parameter on methods it will fail

without a version it should default to 6 but if you send a version in the url it will use that version until you send another one. (so you could send a ping with the version on load and the session will be created with the version.)

This should now cover whatever you were caching and allows people to use the urls  on load as they'll ping the site and the session is recreated.

**Comment by izyspania** [comment]
Thank you man , you put a lot of work into it i see and you probably had other things more important to do at the moment, what i was caching are the stream links from the current album(book),each book has multiple chapters(songs)  and when you open a chapter it adds all the album chapter(songs) in the a que then plays the track you selected  so for example if you i put the  app in the background or i pause the current track for a long time and the session will expire i wont be able to resume the current track nor play any other tracks in the que , the que is generated only when you play a track first time from an album , i was working on recreating the links with the new auth after getting a new session auth (making the stream links dynamic), i created the app at start with subsonic API in mind and its own limitations and using the ampache API complicated a bit things for me as i had to rewrite some code and the logic behind it.

Also some parts of the app is loading multiple queries in parallel and if the key expired then it will ask the server for a new key multiple times (and ill get different keys) for each query and i was not sure how to handle that as im doing the re-auth after the query response is "session expired", for most part of the app im waiting for the query response before sending another request but on things like the "Home Page" i have to do them in parallel ,i can make everything await and but it gets slower on loading which i dont want. I was thinking just to stop all queries when the session expires and display a message that the session expired and reload the app, a bit inconvenient and i have to see how i can deal in case the user only uses the player from the notification area, there are solutions but they all are a bit inconvenient.

I wasnt really working on auth part now but ill do it in the next few days and ill see what solution will be better, ill check out the http-header-authentication too.


![image](https://github.com/ampache/ampache/assets/93366907/cd616da1-050d-4c48-9486-c0cac6f75556)
![image](https://github.com/ampache/ampache/assets/93366907/bab276de-da10-4716-9b7d-7363c63e3bbb)
![image](https://github.com/ampache/ampache/assets/93366907/9016af97-06ce-41d2-8a10-5c85df5f1aec)

**Comment by lachlan-00** [comment]
streams have a few more options for url parameters that will never expire
![image](https://github.com/ampache/ampache/assets/1305249/9bd85bc5-5f23-40cf-b143-aae7950850a7)

'apikey' should be using your handshake auth key (https://ampache.org/api/#api-key). That will authenticate you every time you stream.

```
http://music.com/play/index.php?apikey=authparameterkeyhere&type=song&oid=1094201&uid=1&player=api&name=Christoph%20Gray%20-%20World%201.mp3
```

stream_tokens will also never expire when set as the ssid (user_create_streamtoken in config file)

We can also hack the sessions to never expire. I'm going to add a system option "API sessions never expire"

that way when you make a session it will idealy expire without a goodbye. that's probably the easiest way to do it. i'll add the ability to clear those sessions from the server too.

You can then use those sessions within the http header to hide them from the url because if it doesn't expire someone could see it and use it
